### PR TITLE
Close channels on Timeout

### LINKED ISSUE #294: Automating with zbchaos fails with time outs on gateway during deploying processes
After restructuring the zbchaos and running against our clusters, it often fails with timeout errors on deploying a process model.

This happened when automating with Testbench, running against SaaS, but also when executing locally and running against a self-managed cluster.

```
{"level":"debug","logging.googleapis.com/labels":{"clusterId":"","jobKey":"2251799813688688","processInstanceKey":"2251799813688668","title":"CPU stress on an Broker"},"logging.googleapis.com/operation":{"id":"2251799813688688"},"time":"2022-12-10T14:18:30.854571594+01:00","message":"Port forward to zell-chaos-zeebe-gateway-7796d8579f-pdttw"}
{"level":"debug","logging.googleapis.com/labels":{"clusterId":"","jobKey":"2251799813688688","processInstanceKey":"2251799813688668","title":"CPU stress on an Broker"},"logging.googleapis.com/operation":{"id":"2251799813688688"},"time":"2022-12-10T14:18:30.955492888+01:00","message":"Successfully created port forwarding tunnel"}
{"level":"debug","logging.googleapis.com/labels":{"clusterId":"","jobKey":"2251799813688688","processInstanceKey":"2251799813688668","title":"CPU stress on an Broker"},"logging.googleapis.com/operation":{"id":"2251799813688688"},"time":"2022-12-10T14:18:30.955897778+01:00","message":"Deploy file bpmn/one_task.bpmn (size: 2526 bytes)."}
panic: rpc error: code = DeadlineExceeded desc = Time out between gateway and broker: Request ProtocolRequest{id=3029, subject=command-api-1, sender=10.0.5.86:26502, payload=byte[]{length=2675, hash=1668380772}} to zell-chaos-zeebe-0.zell-chaos-zeebe.zell-chaos.svc:26501 timed out in PT15S
```

Interesting is that when this happens with Testbench and the deployed zbchaos worker the worker retries every 5 minutes (after job timeout) and fails again until the experiment fails due to timeout. I was not able to deploy either by hand with zbctl, which is a bit suspicious to me. It looks like that there is an issue in the gateway or broker when running such experiments.

### Details

As mentioned above I have run the experiment locally, against an self-managed cluster and increased the log level (Zeebe Trace, Atomic Debug). 

When the experiment fail then we can see in the gateway, that the request is sent to the broker:

```

2022-12-10 14:18:30.940 CET
￼
zeebe-gateway
Send request 1519360045 to zell-chaos-zeebe-0.zell-chaos-zeebe.zell-chaos.svc:26501 with topic command-api-1
```

(At least I hope this is the request ?!)

In the broker I see several issues polling etc between leader and followers and later having different leaders.

```
2 - Failed to probe Member{id=0e1d070c-36bb-4023-8b6c-706186e68289, address=zell-chaos-zeebe-0.zell-chaos-zeebe.zell-chaos.svc:26502, properties={}, version=null, timestamp=0, state=null, incarnationNumber=0}
```

Furthermore I see that an response can't be sent back (from broker 0, which is our leader at this time):

```
2022-12-10 14:18:30.654 CET

zell-chaos-zeebe-0

zeebe
Received a reply for message id:[2134] but was unable to locate the request handle
```

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description
We have seen in several experiments that channels have been reused even after a long period of unavailability of the other side. Meaning that the brokers were already restarted and had new IPs but gateways still have sent requests to old IPs without realizing the issue. The problem here is that we have implemented the timeout on top because as far as we can see right now the channels don't provide that feature. The channel response handling was never called, which means no action could be taken here. 

Our fix is now to listen for the timeout exception in order to close the channel. This will cause us to create a new channel on the next request, which allows us to reconnect to the correct node/IP. Please see related investigation https://github.com/zeebe-io/zeebe-chaos/issues/294

I tried to add two new test cases, one which verifies that the channel is closed on timeout and another one that verifies that a new channel is opened after the timeout and the next request. The NettMessagingService is right now hard to test and has a lot of internals, which we need for our tests which causes to be our tests not that clean as I would like to have them. I hope we can improve soon on this code part or replace them with grpc.

I will run some benchmarks with this branch to see whether it really solves this issue, but a previous QA run succeeded :crossed_fingers: 

## Related issues

closes https://github.com/zeebe-io/zeebe-chaos/issues/294

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by Zelldon** [comment]
QA run succeeded https://camunda.slack.com/archives/C013MEVQ4M9/p1671552604143189 ❤

File: atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java [hunk 1]
```diff
@@ -130,9 +130,27 @@ public NettyMessagingService(
     this.advertisedAddress = advertisedAddress;
     this.protocolVersion = protocolVersion;
     this.config = config;
+    this.channelPool = new ChannelPool(this::openChannel, config.getConnectionPoolSize());
+
+    openFutures = new CopyOnWriteArrayList<>();
+    initAddresses(config);
+  }
+
+  // duplicated for tests - to inject channel pool
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java:139]:
```suggestion
  @VisibleForTesting
```

We have this annotation as part of zeebe-util. We could extend it to add a reason/justification as well :+1:

**Review by npepinpe — APPROVED** [review]
- :wrench: I would be interested in differentiating protocol/application errors from network errors, so we can simply close the channel whenever any unexpected error occurs. But that's more work, and we should merge this anyway as there's immediate value to be gained :+1: 
- :+1: Glad to see tests! You're right that it's awkward to test, and we could refactor it to make it simpler, or invest into moving away from it to a higher level abstraction (e.g. gRPC). We can discuss that in a team meeting and decide on a time frame (e.g. when should we start)

**Comment by npepinpe** [comment]
I think we should backport this as well :+1:

**Comment by Zelldon** [comment]
Thanks for the review @npepinpe :bow: 

bors r+

**Pull request merged**